src/python/pants/backend/python/rules/black_test.py

```diff
@@ -0,0 +1,138 @@
+# Copyright 2019 Pants project contributors (see CONTRIBUTORS.md).
+# Licensed under the Apache License, Version 2.0 (see LICENSE).
+
+from typing import List, Optional, Tuple
+
+from pants.backend.python.rules.black import BlackSetup, fmt, lint, setup_black
+from pants.backend.python.rules.download_pex_bin import download_pex_bin
+from pants.backend.python.rules.pex import CreatePex, create_pex
+from pants.backend.python.subsystems.black import Black
+from pants.backend.python.subsystems.python_native_code import (
+  PythonNativeCode,
+  create_pex_native_build_environment,
+)
+from pants.backend.python.subsystems.python_setup import PythonSetup
+from pants.backend.python.subsystems.subprocess_environment import (
+  SubprocessEnvironment,
+  create_subprocess_encoding_environment,
+)
+from pants.backend.python.targets.formattable_python_target import FormattablePythonTarget
+from pants.build_graph.address import Address
+from pants.engine.fs import Digest, FileContent, InputFilesContent, Snapshot
+from pants.engine.legacy.structs import TargetAdaptor
+from pants.engine.rules import RootRule
+from pants.engine.selectors import Params
+from pants.rules.core.fmt import FmtResult
+from pants.rules.core.lint import LintResult
+from pants.source.wrapped_globs import EagerFilesetWithSpec
+from pants.testutil.subsystem.util import global_subsystem_instance, init_subsystems
+from pants.testutil.test_base import TestBase
+from pants.util.collections import assert_single_element
+
+
+class TestBlack(TestBase):
+
+  good_source = FileContent(path="test/good.py", content=b'name = "Anakin"\n')
+  bad_source = FileContent(path="test/bad.py", content=b'name=    "Anakin"\n')
+  fixed_bad_source = FileContent(path="test/bad.py", content=b'name = "Anakin"\n')
+
+  @classmethod
+  def rules(cls):
+    return (
+      *super().rules(),
```

Comment:
NB: This is still an integration test! It's a notch better than via the CLI though.

Reply:
Note further - if individual rules are actuall unit-tested, the importance of an integration test is drastically reduced in any pure function setting.

Reply:
Hm, true. Should we be naming these tests `foo_integration_test.py` and tagging them with `'integration'`? I think that it could be meaningful to still distinguish and will encourage more unit tests via `run_rule()`. Timeouts are also much higher for these than `run_rule()`-based tests.

Reply:
I think that sounds like a great idea, because the distinction isn't as clear now as it was with v1 tasks!

Reply:
Probably, yes. To repeat the xoogler sin: one way of doing this is categorizing size. This could be manual - small/medium/large, with automatic timeouts capping each, say 100ms, 1s, >1s. You can also pair sandboxing with this - a small test should never do IO and so you run it in a sandbox where IO primitves raise.

Reply:
Noting we can acheive the latter in a language agnostic way with remoting by turning off network access.

Reply:
Google uses starlark for many integration tests internally actually, which is similar in many ways to the v2 engine's python API, so maybe not the worst analogy.

> You can also pair sandboxing with this - a small test should never do IO and so you run it in a sandbox where IO primitves raise.

I like this idea a lot.

Reply:
Hm... so, this is extending `TestBase`, meaning that it does not consume any external anything: it has empty/default config with a new anonymous workdir and buildroot and etc... the inputs could be "in memory" for all that the test actually cares.

In our current definitions of "unit" vs "integration", this is a unit test. But it's pretty clearly a spectrum, and v2 rules create a new spectrum.

Reply:
> In our current definitions of "unit" vs "integration", this is a unit test. But it's pretty clearly a spectrum, and v2 rules have a new spectrum.

Is the only purpose of this distinction for sharding CI? If so, the distinction since less relevant with remoting.

cc @benjyw on this whole thread.